### PR TITLE
Move scheduler loop into AP_Scheduler

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

**Comment by peterbarker** [comment]
The most significant change functionality-wise in this PR is that it fixes the `PERF` reporting of all vehicles.

As it stands, the time we spend "waiting for an ins sample" (*cough*) is counted as part of the loop time, which is not great from a utility point of view.

This is what the timings look like after this patch (when running the main loop at 1kHz):

```
APM: PERF: 244/10000 max=2121 min=290 avg=663 sd=218
APM: PERF: 247/10000 max=2137 min=287 avg=666 sd=220
APM: PERF: 248/10002 max=2207 min=290 avg=666 sd=221
APM: PERF: 242/10000 max=2220 min=284 avg=668 sd=221
APM: PERF: 242/10000 max=2225 min=284 avg=666 sd=221
APM: PERF: 247/10000 max=2075 min=283 avg=663 sd=222
APM: PERF: 247/10000 max=2227 min=284 avg=667 sd=221
APM: PERF: 243/10000 max=2192 min=278 avg=666 sd=220
APM: PERF: 241/10000 max=2140 min=288 avg=668 sd=218
APM: PERF: 253/10000 max=2109 min=283 avg=666 sd=219
APM: PERF: 250/10000 max=2206 min=288 avg=667 sd=220
APM: PERF: 246/10000 max=2187 min=285 avg=665 sd=223
APM: PERF: 247/10000 max=2148 min=284 avg=665 sd=224
APM: PERF: 246/10000 max=2104 min=283 avg=665 sd=219
APM: PERF: 249/10000 max=2175 min=287 avg=664 sd=222
APM: PERF: 241/10001 max=2152 min=287 avg=669 sd=221
APM: PERF: 237/10000 max=2162 min=281 avg=663 sd=222
APM: PERF: 245/10000 max=2183 min=288 avg=663 sd=223
APM: PERF: 246/10000 max=2073 min=288 avg=668 sd=219
APM: PERF: 257/10000 max=2150 min=283 avg=664 sd=220
APM: PERF: 248/10000 max=2209 min=287 avg=664 sd=222
APM: PERF: 253/10000 max=2153 min=284 avg=663 sd=223
```

**Comment by peterbarker** [comment]
Speculation welcome on how, when we're targeting a 1000us loop we can over-run by 1000us!  Measurements even more welcome :-)

The measurements above were taken on a PixRacer.

**Comment by khancyr** [comment]
looks pretty good on first read. It will take me more read time to look at changes carrefully !

**Comment by peterbarker** [comment]
On Wed, 15 Nov 2017, Francisco Ferreira wrote:

> I like the changes, but I hoped it had gone further:
>
>  *  make the performance log message common

Yep, I crossed this one with tridge, he's happy to have a common format. 
That means removing the gyro drift stuff from Rover!

I'll do some more patches for this.

>  *  integrate Copter and Sub fast loop into the tasks instead of running it
>     separately

It's not a bad option, but worthy of a separate PR, I think.

> Also an idea that crossed my mind while reviewing this was to make vehicles
> inherit from AP_Scheduler (and remove that ugly loop calling loop 🙂).

Bleh, yes, that's nasty, but I do note you didn't instantly suggest a 
better name :-)

That's an interesting idea, inheritting from Scheduler.  I'm  pretty sure 
that would meet with some resistance :-)

> I would also like to see a sizable squash, but we can leave that for later
> discussion.

Yep.

> In ArduCopter/ArduCopter.cpp:
> 
> > @@ -178,9 +178,7 @@ void Copter::setup()
>      scheduler.init(&scheduler_tasks[0], ARRAY_SIZE(scheduler_tasks));
>
>      // setup initial performance counters
> -    perf_info.set_loop_rate(scheduler.get_loop_rate_hz());
> -    perf_info.reset();
> -    fast_loopTimer = AP_HAL::micros();
> +    scheduler.set_loop_rate(scheduler.get_loop_rate_hz());
> 
> Looks strange, why do we need to set the scheduler loop rate from the
> scheduler loop rate?

It does look very odd.  I meant to move that setup stuff into 
scheduler.init, so I'll fix that up.

> In libraries/AP_Scheduler/AP_Scheduler.cpp:
> 
> > @@ -233,6 +230,13 @@ void AP_Scheduler::loop()
>      const uint32_t loop_us = get_loop_period_us();
>      const uint32_t time_available = (timer + loop_us) - AP_HAL::micros();
>      run(time_available > loop_us ? 0u : time_available);
> +
> +    // check loop time
> +#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
> +    // move result of AP_HAL::micros() forward:
> +    hal.scheduler->delay_microseconds(1);
> 
> Interesting.

Indeed.  There may be a lower-level call that could be made there, but 
this works and will have minimal impact.

**Comment by OXINARF** [comment]
> That means removing the gyro drift stuff from Rover!

I have my doubts of the need for that information, but we can do another message just for it if Rover people still want it. Having it in the performance message seems a bit strange to me.

> Bleh, yes, that's nasty, but I do note you didn't instantly suggest a
better name :-)

I don't think there is a good name for it. It's the approach that is bad, not the name. Just to be clear, I understand why you went that path, it's me that hate it just as much as I hate all the vehicle methods that just call library updates.

> That's an interesting idea, inheritting from Scheduler.  I'm  pretty sure
that would meet with some resistance :-)

Probably, but makes perfect sense to me. Either that or the scheduler "owning" the vehicle. What is running on top of the HAL should be the scheduler and the scheduler is running several tasks, one of which is the vehicle code 🙂
The loop rate being coordinated by the InertialSensor library is another thing that drives me crazy, but well... I'll fix the world in another life 😛

**Comment by OXINARF** [comment]
> It's not a bad option, but worthy of a separate PR, I think.

Forgot this one. I won't oppose doing it in a separate PR, but doing it here would simplify this PR.

File: libraries/AP_Scheduler/AP_Scheduler.cpp [hunk 1]
```diff
@@ -233,6 +230,13 @@ void AP_Scheduler::loop()
     const uint32_t loop_us = get_loop_period_us();
     const uint32_t time_available = (timer + loop_us) - AP_HAL::micros();
     run(time_available > loop_us ? 0u : time_available);
+
+    // check loop time
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+    // move result of AP_HAL::micros() forward:
+    hal.scheduler->delay_microseconds(1);
```
[the remark this answers is not in the document]
Comment (hiro2233) [reply libraries/AP_Scheduler/AP_Scheduler.cpp]: try to call it internally on the tick() and delete the name "loop" on the scheduler, it's confusing to the main loop.

**Comment by peterbarker** [comment]
On Thu, 16 Nov 2017, Francisco Ferreira wrote:

>       That's an interesting idea, inheritting from Scheduler. I'm
>       pretty sure
>       that would meet with some resistance :-)
> 
> Probably, but makes perfect sense to me. Either that or the scheduler
> "owning" the vehicle. What is running on top of the HAL should be the
> scheduler and the scheduler is running several tasks, one of which is the
> vehicle code 🙂
> The loop rate being coordinated by the InertialSensor library is another
> thing that drives me crazy, but well... I'll fix the world in another life
> 😛

There's another couple of options:
  - AP_Scheduler_Copter, AP_Scheduler_Plane, ...
  - do what DataFlash (sorry, AP::Logging) does with its log structures, 
using #define tricks to create a monster structure based on both library 
and vehicle-specific data.

Both of those suffer from the problem that each vehicle will want 
different default rates for library "update()"s - and that's pretty ugly 
where we currently do it.

... and... we still don't have a good solution for scaling frequency up 
and down.

Say we increase the loop rate to 1kHz in Copter.  All of the "fast" 
scheduling entries still say 400Hz, so any advantage is significantly 
reduced.  Don't really just want to multiply stuff out - not much point 
navigating faster than we currently do, for example.

That, by the way, is an argument for not doing away with the Copter and 
Sub's "fast loop" for the time being!  (I'm ignoring the wait_for_sample() 
being based on the loop rate for the purposes of this argument as a 
simple inconvenience....)

**Comment by peterbarker** [comment]
@OXINARF @khancyr I've removed WIP from this after doing a bit of squashing.

@hiro2233 Finally understood your comment, I think!  Yes, I thought about moving this to be part of `tick()` - but that can wait, I reckon.

File: APMrover2/APMrover2.cpp [hunk 2]
```diff
@@ -78,6 +78,7 @@ const AP_Scheduler::Task Rover::scheduler_tasks[] = {
     SCHED_TASK(accel_cal_update,       10,    100),
     SCHED_TASK(dataflash_periodic,     50,    300),
     SCHED_TASK(ins_periodic,           50,     50),
+    SCHED_TASK(perf_update,           0.1,     75),
```
Comment (khancyr) [inline APMrover2/APMrover2.cpp]: 0.1f

**Comment by peterbarker** [comment]
On Fri, 1 Dec 2017, Pierre Kancir wrote:

>      SCHED_TASK(accel_cal_update,       10,    100),
>      SCHED_TASK(dataflash_periodic,     50,    300),
>      SCHED_TASK(ins_periodic,           50,     50),
> +    SCHED_TASK(perf_update,           0.1,     75),
> 
> 0.1f

Ordinarily, maybe, but (a) there's no room in the structure for anything 
but a float (so no type promotion), and (b) What Would Tridge Do? (omit 
the 'f' as it turns out ;-) )

File: libraries/AP_Scheduler/AP_Scheduler.cpp [hunk 3]
```diff
@@ -198,3 +206,66 @@ float AP_Scheduler::load_average() const
...
+
+void AP_Scheduler::loop()
+{
+    // wait for an INS sample
+    AP::ins().wait_for_sample();
+
+    const uint32_t timer = AP_HAL::micros();
+
+    // used by PI Loops
+    G_Dt                    = (float)(timer - loop_start) / 1000000.0f;
```
Comment (khancyr) [inline libraries/AP_Scheduler/AP_Scheduler.cpp]: I am a little mixed about the large division but * 0.000001f doesn't look much better

File: libraries/AP_Scheduler/AP_Scheduler.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (khancyr) [reply libraries/AP_Scheduler/AP_Scheduler.cpp]: I don't understand why having a delay in SITL ?

File: libraries/AP_Scheduler/PerfInfo.cpp [hunk 4]
```diff
@@ -94,3 +95,15 @@ uint32_t AP::PerfInfo::get_stddev_time() const
 {
     return sqrt((sigmasquared_time - (sigma_time*sigma_time)/loop_count) / loop_count);
 }
+
+void AP::PerfInfo::update_logging()
```
Comment (khancyr) [inline libraries/AP_Scheduler/PerfInfo.cpp]: maybe more report_logging ? or send_logging ? I don't know.... or stay as it is also good enough

File: libraries/AP_Scheduler/PerfInfo.h [hunk 5]
```diff
@@ -24,10 +26,18 @@ class PerfInfo {
     uint32_t get_avg_time() const;
     uint32_t get_stddev_time() const;
 
+    void set_loop_rate(uint16_t rate_hz) {
+        // allow a 20% overrun before we consider a loop "slow":
+        overtime_threshold_micros = 1000000/rate_hz * 1.2;
```
Comment (khancyr) [inline libraries/AP_Scheduler/PerfInfo.h]: I would say 1.2f ;-P

**Comment by peterbarker** [comment]
On Fri, 1 Dec 2017, Pierre Kancir wrote:

> I don't understand why having a delay in SITL ?

If you don't do some sort of a delay then micros() doesn't move; lock-step 
scheduling.

**Comment by khancyr** [comment]
Ok LTGM

**Comment by peterbarker** [comment]
On Fri, 1 Dec 2017, Pierre Kancir wrote:

> +        overtime_threshold_micros = 1000000/rate_hz * 1.2;
> I would say 1.2f ;-P

Fixed, thanks!

**Comment by peterbarker** [comment]
On Fri, 1 Dec 2017, Pierre Kancir wrote:

> Ok LTGM

Thanks for the review!

**Comment by khancyr** [comment]
@peterbarker maybe add your comment as comment on the delay for SITL would be a good thing!

**Comment by peterbarker** [comment]
On Fri, 1 Dec 2017, Pierre Kancir wrote:

> @peterbarker maybe add your comment as comment on the delay for SITL would
> be a good thing!

It's already there?! :-)

**Comment by khancyr** [comment]
yep my bad ...

**Comment by hiro2233** [comment]
@peterbarker I have running Scheduler without waiting the sample, full free running and no delays, the sensor data consumption is performed only on the callback.

With the FIFO and IRQ help we doesn't need to wait for samples on the front-end. The goal is not use dalays > Callbacks timers.

And i think that perfomance monitor need to be preloaded on the init and putting it on callback to run in parallel to it at the maximum system speed timer.

```
// one tick has passed
void AP_Scheduler::tick(void)
{
    _time_now = AP_HAL::micros();

    if ((_time_now - _time_last) > get_loop_period_us() - 1)  {
        _tick_counter++;
        _time_last = AP_HAL::micros();
    }
}
```

**Comment by peterbarker** [comment]
On Fri, 1 Dec 2017, Hiroshi Takey (hiro2233) wrote:

> @peterbarker I have running Scheduler without waiting the sample, full free
> running and no delays, the sensor data consumption is performed only on the
> callback.
> 
> With the FIFO and IRQ help we doesn't need to wait for samples on the
> front-end. The goal is not use dalays > Callbacks timers.
> 
> And i think that perfomance monitor need to be preloaded on the init and
> putting it on callback to run in parallel to it at the maximum system speed
> timer.

That's all entirely possible.  Does it need to be done on this PR?

**Comment by hiro2233** [comment]
Yes, but you modified several and critical parts on some libraries like motors, it's risky, the performance inclusion into the scheduler is not a good idea to maintan a synchronization, the preformance monitor need to be in separate thread running synchronous to this scheduler. The only that we need to monitor scheduling is a simple "switch event" just before firing the task. The others libraries doesn't need to know what happen in the scheduler and thats is the rule on real time.

But to reach this we need work with IRQ or ISR and this need be connected directly to the CPU performance monitor counters. On Intel architecture have two fast RT timers running at 180 Mhz - 200 Mhz to do it, this speed is amazing and is the only one that i see theorically, but it is behind the CPU and i asked to Intel to use these registers and told me that these registers can't be used on user space, and nothing else. On ARM it's possible to acces to this CPU perf counters but we need at least two timers available, on AVR have 5 timers that is connected directly to CPU and it's is very good to make it at real time synch but the problem is the limited speed CPU, with this architecture and with more speed we can reach and calculate the speed of photons with less error.

**Comment by khancyr** [comment]
did you just rebase or change something ?

**Comment by peterbarker** [comment]
@khancyr Don't think so...

**Comment by dgrat** [comment]
Starting a loop grabbing sensor data is not really what I would expect from a scheduler. In my opinion the correct way to solve the issue is to make an abstract vehicle class from which all other vehicles get derived. There would be a virtual functions implementing the scheduler loop). Then there is no code copy and the scheduler would be still concise. Am I right?

**Comment by peterbarker** [comment]
On Fri, 12 Jan 2018, Daniel Frenzel wrote:

> Starting a loop grabbing sensor data is not really what I would expect from a scheduler. In my opinion the correct way to solve the issue is to make an
> abstract vehicle class from which all other vehicles get derived. There would be a virtual functions implementing the scheduler loop). Then there is no
> code copy and the scheduler would be still concise. Am I right?

Ah, the much-mooted vehicle base class :-)

Considering the scheduler does stuff not actually vehicle-related (the 
user hooks, for example), a separate entiry is sensible IMO.

**Comment by dgrat** [comment]
@peterbarker 
What do you mean with separate entry?

**Comment by peterbarker** [comment]
On Mon, 22 Jan 2018, Daniel Frenzel wrote:

> @peterbarker
> What do you mean with separate entry?

Sorry, separate *entity*.

Currently we create a top-level "vehicle" object and the scheduler 
*object* goes in there.  That works.

File: libraries/AP_Scheduler/AP_Scheduler.cpp [hunk 6]
```diff
@@ -241,9 +243,28 @@ void AP_Scheduler::loop()
...
+void AP_Scheduler::Log_Write_Performance()
+{
+    struct log_Performance pkt = {
+        LOG_PACKET_HEADER_INIT(LOG_PERFORMANCE_MSG),
+        time_us          : AP_HAL::micros64(),
+        num_long_running : perf_info.get_num_long_running(),
+        num_loops        : perf_info.get_num_loops(),
+        max_time         : perf_info.get_max_time(),
+        ins_error_count  : AP::ins().error_count(),
+        hal.util->available_memory()
```
Comment (OXINARF) [inline libraries/AP_Scheduler/AP_Scheduler.cpp]: Adding the var name to this one would be nice.

File: libraries/DataFlash/LogStructure.h [hunk 7]
```diff
@@ -1217,6 +1227,8 @@ Format characters in the format string for binary log messages
       "BCN", "QBBfffffff",  "TimeUS,Health,Cnt,D0,D1,D2,D3,PosX,PosY,PosZ", "s--mmmmmmm", "F--BBBBBBB" }, \
     { LOG_PROXIMITY_MSG, sizeof(log_Proximity), \
       "PRX", "QBfffffffffff", "TimeUS,Health,D0,D45,D90,D135,D180,D225,D270,D315,DUp,CAn,CDis", "s-mmmmmmmmmhm", "F-BBBBBBBBB00" }, \
+    { LOG_PERFORMANCE_MSG, sizeof(log_Performance),                     \
+            "PM",  "QHHIHI",    "TimeUS,NLon,NLoop,MaxT,INSErr,Mem", "s----b", "F----0" }, \
```
Comment (OXINARF) [inline libraries/DataFlash/LogStructure.h]: Too many spaces?

File: libraries/AP_Scheduler/AP_Scheduler.cpp [hunk 8]
```diff
@@ -197,3 +205,66 @@ float AP_Scheduler::load_average()
...
+
+    // check loop time
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+    // move result of AP_HAL::micros() forward:
+    hal.scheduler->delay_microseconds(1);
+#endif
+    perf_info.check_loop_time(AP_HAL::micros() - loop_start);
+}
+
+void AP_Scheduler::update_logging(const bool log_to_dataflash)
```
Comment (OXINARF) [inline libraries/AP_Scheduler/AP_Scheduler.cpp]: Should it get the bit instead (like you've done for other libraries)?

**Comment by peterbarker** [comment]
On Fri, 9 Feb 2018, Francisco Ferreira wrote:

> Made a couple more comments but in general looks OK. A couple more squash
> would be nice, but I won't hold the PR on it.

Cool.

> It would also be nice if vehicles didn't need to call update_logging,
> scheduler should have all info it needs to do it for itself. But I guess
> that can stay for later improvement.

Yep.

> I just realized that the loop time check is changed, so I think this should
> pass by the dev call to see if everyone is OK with it; I'll add the label
> but I'll probably not be at the next dev call.

The loop time change is a bug fix AFAICS; we wouldn't count the last task.

Sadly, your review crossed over with a rebase on top of tridge's work, so 
some of your comments are outdated.

> -    if( time_in_micros > OVERTIME_THRESHOLD_MICROS ) {
> Can't we remove the macro definition?

Already gone.

>      // in multiples of the main loop tick. So if they don't run on
>      // the first call to the scheduler they won't run on a later
>      // call until scheduler.tick() is called again
> -    scheduler.run(loop_us);
> +    const uint32_t loop_us = scheduler.get_loop_period_us();
> +    const uint32_t time_available = (timer + loop_us) - micros();
> +    scheduler.run(time_available);
> 
> Should use the same fix that was done in Copter before:
> 
> https://github.com/ArduPilot/ardupilot/blob/33be8de1f374c1955eab5b1ad8a753f
> e44184517/ArduCopter/ArduCopter.cpp#L241

That check's been moved into AP_Scheduler:

https://github.com/peterbarker/ardupilot/blob/scheduler-loop-move/libraries/AP_Scheduler/AP_Scheduler.cpp#L234

> +#include <AP_InertialSensor/AP_InertialSensor.h>
> 
> Nitpicking: can we include this just in the cpp file? Reduces scope.

Yep.  Still compiles :-)


> In libraries/AP_Scheduler/AP_Scheduler.h:
> 
> > @@ -119,7 +128,16 @@ class AP_Scheduler
>      // current running task, or -1 if none. Used to debug stuck tasks
>      static int8_t current_task;
> 
> +    // loop performance monitoring:
> +    AP::PerfInfo perf_info;
> +
> +    // time taken for previous loop:
> +    float G_Dt;
> 
> Nitpicking: since we are moving this, can't we give it a better name? I
> never remember that G_Dt is the last real loop time.

Will do.

> In libraries/AP_Scheduler/examples/Scheduler_test/Scheduler_test.cpp:
> 
> > @@ -17,7 +17,7 @@ class SchedTest {
>  private:
>
>      AP_InertialSensor ins;
> -    AP_Scheduler scheduler;
> +    AP_Scheduler scheduler{nullptr};
> 
> I can't comment in the appropriate place. Shouldn't the loop method below be
> replaced to only call the scheduler loop?

Yep.  Fixed.

> 
> In libraries/AP_Scheduler/AP_Scheduler.cpp:
> 
> > +        Log_Write_Performance();
> +    }
> +    perf_info.reset();
> +}
> +
> +// Write a performance monitoring packet
> +void AP_Scheduler::Log_Write_Performance()
> +{
> +    struct log_Performance pkt = {
> +        LOG_PACKET_HEADER_INIT(LOG_PERFORMANCE_MSG),
> +        time_us          : AP_HAL::micros64(),
> +        num_long_running : perf_info.get_num_long_running(),
> +        num_loops        : perf_info.get_num_loops(),
> +        max_time         : perf_info.get_max_time(),
> +        ins_error_count  : AP::ins().error_count(),
> +        hal.util->available_memory()
> 
> Adding the var name to this one would be nice.

Will do.


> 
> In libraries/DataFlash/LogStructure.h:
> 
> > @@ -1217,6 +1227,8 @@ Format characters in the format string for binary lo
> g messages
>        "BCN", "QBBfffffff",  "TimeUS,Health,Cnt,D0,D1,D2,D3,PosX,PosY,PosZ",
>  "s--mmmmmmm", "F--BBBBBBB" }, \
>      { LOG_PROXIMITY_MSG, sizeof(log_Proximity), \
>        "PRX", "QBfffffffffff", "TimeUS,Health,D0,D45,D90,D135,D180,D225,D270
> ,D315,DUp,CAn,CDis", "s-mmmmmmmmmhm", "F-BBBBBBBBB00" }, \
> +    { LOG_PERFORMANCE_MSG, sizeof(log_Performance),                     \
> +            "PM",  "QHHIHI",    "TimeUS,NLon,NLoop,MaxT,INSErr,Mem", "s----
> b", "F----0" }, \
> 
> Too many spaces?

I'm wondering about reformatting all of that stuff at some stage.

Fixed.

> +void AP_Scheduler::update_logging(const bool log_to_dataflash)
> Should it get the bit instead (like you've done for other libraries)?

Added as a patch on top.

**Comment by tridge** [comment]
rebased on master

**Comment by tridge** [comment]
the rebases passes CI, but it produces completely incorrect scheduling numbers. For example, plane at 400Hz on a pixracer:
APM: PERF: 2000/2000 max=2210 min=253 avg=308 sd=131
APM: PERF: 2000/2000 max=1385 min=253 avg=308 sd=124
APM: PERF: 2000/2000 max=2408 min=253 avg=308 sd=133
APM: PERF: 2000/2000 max=1382 min=253 avg=308 sd=126

thats wrong in several ways:

- number of misses is wrong
- avg time is crazy - no way we are averaging 308us on a pixracer with dual EKF
- max time is crazy

I also think we should probably be using a average loop time for G_Dt, not last loop time. We're trying to predict the next loop time. I know that is a semantic change, but I think it would be correct

**Comment by tridge** [comment]
I've got it fixed up now, and I've updated this PR branch
note that it now uses a filtered loop time, not last loop time. The last loop time calculation was badly incorrect for use in G_Dt

File: APMrover2/failsafe.cpp [hunk 9]
```diff
@@ -16,13 +16,14 @@
  */
 void Rover::failsafe_check()
 {
-    static uint16_t last_mainLoop_count;
+    static uint16_t last_ticks;
```
Comment (khancyr) [inline APMrover2/failsafe.cpp]: failsafe_last_ticks to be even with other vehicle

File: ArduCopter/mode.h [hunk 10]
```diff
@@ -130,7 +130,7 @@ class ModeAcro_Heli : public ModeAcro {
 
 public:
     // inherit constructor
-    using Copter::Mode::Mode;
+    using Copter::ModeAcro::Mode;
```
Comment (khancyr) [inline ArduCopter/mode.h]: this should be move out in another PR

File: ArduPlane/failsafe.cpp [hunk 11]
```diff
@@ -16,14 +16,15 @@
  */
 void Plane::failsafe_check(void)
 {
-    static uint16_t last_mainLoop_count;
+    static uint16_t last_ticks;
```
Comment (khancyr) [inline ArduPlane/failsafe.cpp]: failsafe_last_ticks same as for rover

File: libraries/AP_InertialSensor/AP_InertialSensor.cpp [hunk 12]
```diff
@@ -1995,3 +1995,13 @@ MAV_RESULT AP_InertialSensor::simple_accel_cal(AP_AHRS &ahrs)
 
     return result;
 }
+
+
+namespace AP {
```
Comment (khancyr) [inline libraries/AP_InertialSensor/AP_InertialSensor.cpp]: should be move out

File: libraries/AP_Scheduler/AP_Scheduler.cpp [hunk 13]
```diff
@@ -70,13 +74,19 @@ AP_Scheduler::AP_Scheduler(void)
 }
 
 // initialise the scheduler
-void AP_Scheduler::init(const AP_Scheduler::Task *tasks, uint8_t num_tasks)
+    void AP_Scheduler::init(const AP_Scheduler::Task *tasks, uint8_t num_tasks, uint32_t log_performance_bit)
```
Comment (khancyr) [inline libraries/AP_Scheduler/AP_Scheduler.cpp]: whitespace ?

File: libraries/AP_Scheduler/PerfInfo.cpp [hunk 14]
```diff
@@ -45,11 +46,28 @@ void AP::PerfInfo::check_loop_time(uint32_t time_in_micros)
...
+       predicted time for the next loop. We remove really excessive
+       times from this calculation so as not to throw it off too far
+       in case we get a single long loop
+
+       Note that the time we use here is the time between calls to
+       check_loop_time() not the time from loop start to loop
+       end. This is because we are using the time for time between
+       calls to controllers, which has nothing to do with cpu speed.
+    */
+    uint32_t now = AP_HAL::micros();
```
Comment (khancyr) [inline libraries/AP_Scheduler/PerfInfo.cpp]: const

File: libraries/AP_Scheduler/PerfInfo.cpp [hunk 15]
```diff
@@ -45,11 +46,28 @@ void AP::PerfInfo::check_loop_time(uint32_t time_in_micros)
...
+    uint32_t loop_time_us = now - last_check_us;
```
Comment (khancyr) [inline libraries/AP_Scheduler/PerfInfo.cpp]: const

File: libraries/AP_Scheduler/PerfInfo.cpp [hunk 16]
```diff
@@ -45,11 +46,28 @@ void AP::PerfInfo::check_loop_time(uint32_t time_in_micros)
...
+    last_check_us = now;
+    if (loop_time_us < overtime_threshold_micros + 10000UL) {
+        filtered_loop_time = 0.99 * filtered_loop_time + 0.01 * loop_time_us * 1.0e-6;
```
Comment (khancyr) [inline libraries/AP_Scheduler/PerfInfo.cpp]: f qualifier

File: libraries/AP_Scheduler/PerfInfo.cpp [hunk 17]
```diff
@@ -93,3 +111,32 @@ uint32_t AP::PerfInfo::get_stddev_time() const
...
+}
+
+void AP::PerfInfo::set_loop_rate(uint16_t rate_hz)
+{
+    // allow a 20% overrun before we consider a loop "slow":
+    overtime_threshold_micros = 1000000/rate_hz * 1.2f;
+
+    if (loop_rate_hz != rate_hz) {
+        loop_rate_hz = rate_hz;
+        filtered_loop_time = 1.0 / rate_hz;
```
Comment (khancyr) [inline libraries/AP_Scheduler/PerfInfo.cpp]: f

**Comment by peterbarker** [comment]
On Mon, 12 Feb 2018, Pierre Kancir wrote:

> @khancyr commented on this pull request.
> 
> ____________________________________________________________________________
> 
> In APMrover2/failsafe.cpp:
> 
> > @@ -16,13 +16,14 @@
>   */
>  void Rover::failsafe_check()
>  {
> -    static uint16_t last_mainLoop_count;
> +    static uint16_t last_ticks;
> 
> failsafe_last_ticks to be even with other vehicle

It would then be out of kilter with the other variable, "last_timestamp". 
Namespacing should mean we're OK here, and internal consistency is good 
:-)

I think making this stuff common to all vehicles should be the subject of 
a follow-up PR - probably creating an AP_MainLoopFailSafe_Plane etc. 
Plane's code is a heck of a lot more complicated than everyone else's, 
'though!

**Comment by peterbarker** [comment]
On Mon, 12 Feb 2018, Pierre Kancir wrote:

> In ArduCopter/mode.h:
> 
> > @@ -130,7 +130,7 @@ class ModeAcro_Heli : public ModeAcro {
>
>  public:
>      // inherit constructor
> -    using Copter::Mode::Mode;
> +    using Copter::ModeAcro::Mode;
> 
> this should be move out in another PR

Or just cherry-picked in :-)

**Comment by peterbarker** [comment]
On Mon, 12 Feb 2018, Pierre Kancir wrote:

> failsafe_last_ticks same as for rover

Same argument for Rover - this code is the same through all of the 
vehicles, making it common to all (and namespacing the variables into the 
inheritted base class!) is where we should fix this.

**Comment by peterbarker** [comment]
On Mon, 12 Feb 2018, Pierre Kancir wrote:

> +namespace AP {
> 
> should be move out

As a separate PR it would probably be killed as "no user" :-)

**Comment by peterbarker** [comment]
On Mon, 12 Feb 2018, Pierre Kancir wrote:

> whitespace ?

Will fix.

**Comment by peterbarker** [comment]
On Mon, 12 Feb 2018, Pierre Kancir wrote:

> +    uint32_t now = AP_HAL::micros();
> const

Will fix

**Comment by peterbarker** [comment]
On Mon, 12 Feb 2018, Pierre Kancir wrote:

> +    uint32_t loop_time_us = now - last_check_us;
> const

Will fix

**Comment by peterbarker** [comment]
On Mon, 12 Feb 2018, Pierre Kancir wrote:

> +        filtered_loop_time = 0.99 * filtered_loop_time + 0.01 * loop_time_u
> s * 1.0e-6;
> 
> f qualifier

Will fix.

**Comment by peterbarker** [comment]
Do not merge yet.

```
pbarker@bluebottle:~/rc/ardupilot((no branch, bisect started on scheduler-loop-move))$ git bisect bad
fbe8fe47422b0edd6840ace520228233eb16f418 is the first bad commit
commit fbe8fe47422b0edd6840ace520228233eb16f418
Author: Andrew Tridgell <andrew@tridgell.net>
Date:   Mon Feb 12 14:43:04 2018 +1100

    Copter: use get_last_loop_time_s() for G_Dt

:040000 040000 e97a2d439a003369db258204e3ee350898a32415 a8dc9fdb65c3fc79461c9db4ba5c6472e3b681b3 M	ArduCopter
pbarker@bluebottle:~/rc/ardupilot((no branch, bisect started on scheduler-loop-move))$ 
```

Fly to moon:
```
param set GPS_TYPE 0
arm throttle
rc 3 2000
mode land
```
